File: frontend/js/observations.js
// Observations Module
const Observations = {
    currentObservationData: null,
    currentQuestionIndex: 0,
    questions: [],
    answerOptions: [],
    sessionId: null,
    participantId: null,
    workshopId: null,

    /**
     * Get all observations for a workshop
     */
    async getWorkshopObservations(workshopId) {
        try {
            return await api.get(API_ENDPOINTS.WORKSHOP_OBSERVATIONS(workshopId));
        } catch (error) {
            UI.showToast(`Error loading observations: ${error.message}`, 'error');
            throw error;
        }
    },

    /**
     * Get single observation
     */
    async getObservation(observationId) {
        try {
            return await api.get(API_ENDPOINTS.OBSERVATION_DETAIL(observationId));
        } catch (error) {
            UI.showToast(`Error loading observation: ${error.message}`, 'error');
            throw error;
        }
    },

    /**
     * Get all questions
     */
    async getQuestions() {
        try {
            const data = await api.get(API_ENDPOINTS.OBSERVATION_QUESTIONS);
            this.questions = data.questions;
            this.answerOptions = data.answer_options;
            return data;
        } catch (error) {
            UI.showToast(`Error loading questions: ${error.message}`, 'error');
            throw error;
        }
    },

    /**
     * Initialize observation
     */
    async initializeObservation(sessionId, participantId) {
        try {
            const data = await api.post(API_ENDPOINTS.OBSERVATION_INITIALIZE, {
                session_id: sessionId,
                participant_id: participantId
            });

            this.currentObservationData = data.observation_data;
            this.currentQuestionIndex = 0;

            return data;
        } catch (error) {
            UI.showToast(`Error initializing observation: ${error.message}`, 'error');
            throw error;
        }
    },

    /**
     * Save observation
     */
    async saveObservation(freeformNotes = '') {
        try {
            const result = await api.post(API_ENDPOINTS.OBSERVATION_SAVE, {
                observation_data: this.currentObservationData,
                freeform_notes: freeformNotes
            });

            UI.showToast('Observation saved successfully', 'success');
            this.currentObservationData = null;
            this.currentQuestionIndex = 0;
            return result;
        } catch (error) {
            UI.showToast(`Error saving observation: ${error.message}`, 'error');
            throw error;
        }
    },

    /**
     * Delete observation
     */
    async deleteObservation(observationId) {
        try {
            const result = await api.delete(API_ENDPOINTS.OBSERVATION_DETAIL(observationId));
            UI.showToast('Observation deleted successfully', 'success');
            return result;
        } catch (error) {
            UI.showToast(`Error deleting observation: ${error.message}`, 'error');
            throw error;
        }
    },

    /**
     * Start observation workflow
     */
    async startObservation(sessionId, participantId, workshopId) {
        try {
            this.sessionId = sessionId;
            this.participantId = participantId;
            this.workshopId = workshopId;

            // Load questions if not already loaded
            if (this.questions.length === 0) {
                await this.getQuestions();
            }

            // Initialize observation
            const data = await this.initializeObservation(sessionId, participantId);

            // Show observation page
            UI.showPage('observationPage');

            // Render first question
            this.renderQuestion(0);

            // Setup event listeners
            this.setupEventListeners();
        } catch (error) {
            console.error('Error starting observation:', error);
        }
    },

    /**
     * Setup event listeners
     */
    setupEventListeners() {
        // Cancel button
        const cancelBtn = document.getElementById('cancelObservationBtn');
        if (cancelBtn) {
            cancelBtn.onclick = () => this.cancelObservation();
        }

        // Previous button
        const prevBtn = document.getElementById('prevQuestionBtn');
        if (prevBtn) {
            prevBtn.onclick = () => this.previousQuestion();
        }

        // Next button
        const nextBtn = document.getElementById('nextQuestionBtn');
        if (nextBtn) {
            nextBtn.onclick = () => this.nextQuestion();
        }

        // Finish button
        const finishBtn = document.getElementById('finishObservationBtn');
        if (finishBtn) {
            finishBtn.onclick = () => this.finishObservation();
        }
    },

    /**
     * Render question
     */
    renderQuestion(index) {
        if (index < 0 || index >= this.questions.length) return;

        this.currentQuestionIndex = index;
        const question = this.questions[index];

        // Update progress
        const progress = ((index + 1) / this.questions.length) * 100;
        document.getElementById('observationProgress').style.width = `${progress}%`;
        document.getElementById('observationProgressText').textContent =
            `Pregunta ${index + 1} de ${this.questions.length}`;

        // Update category
        document.getElementById('observationCategory').textContent = question.category;

        // Update question text
        document.getElementById('observationQuestion').textContent = question.text;

        // Render answer buttons
        const answersContainer = document.getElementById('observationAnswers');
        const currentAnswer = this.currentObservationData.answers[question.id];

        // Answer option labels in Spanish
        const answerLabels = {
            'yes': 'Sí',
            'no': 'No',
            'partially': 'Parcialmente',
            'not_applicable': 'No Aplica'
        };

        answersContainer.innerHTML = this.answerOptions.map(option => `
            <button class="answer-btn ${currentAnswer === option ? 'selected' : ''}" 
                    data-answer="${option}"
                    onclick="Observations.selectAnswer('${question.id}', '${option}')">
                ${answerLabels[option] || option}
            </button>
        `).join('');

        // Update navigation buttons
        document.getElementById('prevQuestionBtn').disabled = index === 0;
        document.getElementById('nextQuestionBtn').disabled = !currentAnswer;

        // Show/hide finish button
        const isLastQuestion = index === this.questions.length - 1;
        document.getElementById('nextQuestionBtn').style.display = isLastQuestion ? 'none' : 'flex';
        document.getElementById('finishObservationBtn').style.display = isLastQuestion && currentAnswer ? 'flex' : 'none';

        // Add animation
        const card = document.getElementById('observationPage').querySelector('.observation-card');
        card.classList.add('animating');
        setTimeout(() => card.classList.remove('animating'), 200);
    },

    /**
     * Select answer
     */
    selectAnswer(questionId, answer) {
        // Update answer in observation data
        this.currentObservationData.answers[questionId] = answer;

        // Update UI
        const buttons = document.querySelectorAll('.answer-btn');
        buttons.forEach(btn => {
            if (btn.dataset.answer === answer) {
                btn.classList.add('selected');
            } else {
                btn.classList.remove('selected');
            }
        });

        // Enable next button
        document.getElementById('nextQuestionBtn').disabled = false;

        // If last question, show finish button
        const isLastQuestion = this.currentQuestionIndex === this.questions.length - 1;
        if (isLastQuestion) {
            document.getElementById('finishObservationBtn').style.display = 'flex';
        }
    },

    /**
     * Previous question
     */
    previousQuestion() {
        if (this.currentQuestionIndex > 0) {
            this.renderQuestion(this.currentQuestionIndex - 1);
        }
    },

    /**
     * Next question
     */
    nextQuestion() {
        if (this.currentQuestionIndex < this.questions.length - 1) {
            this.renderQuestion(this.currentQuestionIndex + 1);
        }
    },

    /**
     * Finish observation
     */
    async finishObservation() {
        // Show notes section
        document.getElementById('observationNotesSection').style.display = 'block';
        document.getElementById('finishObservationBtn').textContent = 'Guardar Observación';
        document.getElementById('finishObservationBtn').onclick = () => this.saveAndComplete();

        // Scroll to notes
        document.getElementById('observationNotesSection').scrollIntoView({ behavior: 'smooth' });
    },

    /**
     * Save and complete
     */
    async saveAndComplete() {
        const notes = document.getElementById('observationNotes').value.trim();

        try {
            await this.saveObservation(notes);

            // Go back to workshop detail
            if (this.workshopId) {
                await Workshops.viewWorkshop(this.workshopId);
            } else {
                UI.showPage('workshopsPage');
                await Workshops.loadWorkshops();
            }
        } catch (error) {
            console.error('Error saving observation:', error);
        }
    },

    /**
     * Cancel observation
     */
    cancelObservation() {
        if (confirm('¿Estás seguro de que deseas cancelar? Se perderá el progreso actual.')) {
            this.currentObservationData = null;
            this.currentQuestionIndex = 0;

            // Go back to workshop detail
            if (this.workshopId) {
                Workshops.viewWorkshop(this.workshopId);
            } else {
                UI.showPage('workshopsPage');
                Workshops.loadWorkshops();
            }
        }
    },

    /**
     * Render observations list
     */
    renderObservationsList(observations, workshopId) {
        const container = document.getElementById('observations-list');
        if (!container) return;

        if (!observations || observations.length === 0) {
            container.innerHTML = `
                <div class="alert alert-info">
                    <i class="bi bi-info-circle me-2"></i>
                    No observations found.
                </div>
            `;
            return;
        }

        const html = observations.map(obs => `
            <div class="card mb-3" data-observation-id="${obs.id}">
                <div class="card-body">
                    <div class="d-flex justify-content-between align-items-start">
                        <div class="flex-grow-1">
                            <h6 class="card-title">
                                Session ${obs.session_id} - Participant ${obs.participant_id}
                                <span class="badge bg-primary ms-2">v${obs.version}</span>
                            </h6>
                            <p class="card-text">
                                <small class="text-muted">
                                    <i class="bi bi-calendar me-1"></i>
                                    ${new Date(obs.created_at).toLocaleString()}
                                </small>
                            </p>
                            ${obs.freeform_notes ? `
                                <p class="card-text">
                                    <small><strong>Notes:</strong> ${this.escapeHtml(obs.freeform_notes)}</small>
                                </p>
                            ` : ''}
                            <p class="card-text">
                                <small class="text-muted">
                                    ${Object.keys(obs.answers).length} questions answered
                                </small>
                            </p>
                        </div>
                        <div class="btn-group" role="group">
                            <button class="btn btn-sm btn-outline-primary" onclick="Observations.viewObservation(${obs.id})">
                                <i class="bi bi-eye"></i>
                            </button>
                            <button class="btn btn-sm btn-outline-danger" onclick="Observations.confirmDeleteObservation(${obs.id}, ${workshopId})">
                                <i class="bi bi-trash"></i>
                            </button>
                        </div>
                    </div>
                </div>
            </div>
        `).join('');

        container.innerHTML = html;
    },

    /**
     * View observation details
     */
    async viewObservation(observationId) {
        try {
            const observation = await this.getObservation(observationId);

            // Show modal with observation details
            const content = `
                <div class="mb-3">
                    <strong>Session:</strong> ${observation.session_id}<br>
                    <strong>Participant:</strong> ${observation.participant_id}<br>
                    <strong>Version:</strong> ${observation.version}<br>
                    <strong>Created:</strong> ${new Date(observation.created_at).toLocaleString()}
                </div>
                ${observation.freeform_notes ? `
                    <div class="mb-3">
                        <strong>Notes:</strong><br>
                        ${this.escapeHtml(observation.freeform_notes)}
                    </div>
                ` : ''}
                <div class="mb-3">
                    <strong>Answers:</strong><br>
                    <div class="table-responsive">
                        <table class="table table-sm">
                            <thead>
                                <tr>
                                    <th>Question</th>
                                    <th>Answer</th>
                                </tr>
                            </thead>
                            <tbody>
                                ${Object.entries(observation.answers).map(([q, a]) => `
                                    <tr>
                                        <td>${this.escapeHtml(q)}</td>
                                        <td><span class="badge bg-secondary">${this.escapeHtml(a)}</span></td>
                                    </tr>
                                `).join('')}
                            </tbody>
                        </table>
                    </div>
                </div>
            `;

            UI.showModal('Observation Details', content, [
                {
                    text: 'Close',
                    class: 'btn-outline',
                    onclick: 'UI.closeModal()'
                }
            ]);
        } catch (error) {
            console.error('Error viewing observation:', error);
        }
    },

    /**
     * Confirm delete observation
     */
    confirmDeleteObservation(observationId, workshopId) {
        if (confirm('Are you sure you want to delete this observation? This action cannot be undone.')) {
            this.deleteObservationAndRefresh(observationId, workshopId);
        }
    },

    /**
     * Delete observation and refresh list
     */
    async deleteObservationAndRefresh(observationId, workshopId) {
        try {
            await this.deleteObservation(observationId);
            const observations = await this.getWorkshopObservations(workshopId);
            this.renderObservationsList(observations, workshopId);
        } catch (error) {
            console.error('Error deleting observation:', error);
        }
    },

    /**
     * Escape HTML to prevent XSS
     */
    escapeHtml(text) {
        const div = document.createElement('div');
        div.textContent = text;
        return div.innerHTML;
    }
};

File: app/api/observations.py
"""Observation API endpoints."""
from flask import Blueprint, request, jsonify
from flask_jwt_extended import get_jwt_identity
from app.api.decorators import jwt_required_api
from app.services.observation_service import ObservationService
from app.models.observation_questions import (
    get_all_questions, get_question_by_index, get_total_question_count,
    ANSWER_OPTIONS, OBSERVATION_CATEGORIES
)

observations_api_bp = Blueprint('observations_api', __name__, url_prefix='/observations')


@observations_api_bp.route('/workshop/<int:workshop_id>', methods=['GET'])
@jwt_required_api
def list_workshop_observations(workshop_id):
    """
    GET /api/v1/observations/workshop/{workshop_id}
    
    Get all observations for a workshop.
    
    Response: [
        {
            "id": 1,
            "session_id": 1,
            "participant_id": 1,
            "version": 1,
            "answers": {...},
            "freeform_notes": "...",
            "created_at": "..."
        }
    ]
    """
    user_id = int(get_jwt_identity())
    observations, error = ObservationService.get_workshop_observations(workshop_id, user_id)
    
    if error:
        return jsonify({'error': error}), 404
    
    # Convert to dict (ObservationalRecord doesn't have to_dict, so we'll create it)
    result = []
    for obs in observations:
        result.append({
            'id': obs.id,
            'session_id': obs.session_id,
            'participant_id': obs.participant_id,
            'version': obs.version,
            'answers': obs.answers,
            'freeform_notes': obs.freeform_notes,
            'created_at': obs.created_at.isoformat() if obs.created_at else None
        })
    
    return jsonify(result), 200


@observations_api_bp.route('/<int:observation_id>', methods=['GET'])
@jwt_required_api
def get_observation(observation_id):
    """
    GET /api/v1/observations/{observation_id}
    
    Get single observation details.
    
    Response: {
        "id": 1,
        "session_id": 1,
        "participant_id": 1,
        "version": 1,
        "answers": {...},
        "freeform_notes": "...",
        "created_at": "..."
    }
    """
    user_id = int(get_jwt_identity())
    observation, error = ObservationService.get_observation(observation_id, user_id)
    
    if error:
        return jsonify({'error': error}), 404
    
    return jsonify({
        'id': observation.id,
        'session_id': observation.session_id,
        'participant_id': observation.participant_id,
        'version': observation.version,
        'answers': observation.answers,
        'freeform_notes': observation.freeform_notes,
        'created_at': observation.created_at.isoformat() if observation.created_at else None
    }), 200


@observations_api_bp.route('/initialize', methods=['POST'])
@jwt_required_api
def initialize_observation():
    """
    POST /api/v1/observations/initialize
    
    Initialize a new observation session.
    
    Request: {
        "session_id": 1,
        "participant_id": 1
    }
    Response: {
        "observation_data": {
            "session_id": 1,
            "participant_id": 1,
            "answers": {...},  // Pre-filled from previous observation if exists
            "current_index": 0,
            "is_redo": false,
            "previous_version": 0
        },
        "first_question": {...},
        "total_questions": 50
    }
    """
    user_id = int(get_jwt_identity())
    data = request.get_json()
    
    if not data:
        return jsonify({'error': 'Invalid request format'}), 400
    
    session_id = data.get('session_id')
    participant_id = data.get('participant_id')
    
    if not session_id or not participant_id:
        return jsonify({'error': 'Session ID and Participant ID are required'}), 400
    
    observation_data, error = ObservationService.initialize_observation(
        session_id,
        participant_id,
        user_id
    )
    
    if error:
        return jsonify({'error': error}), 400
    
    # Get first question
    first_question = get_question_by_index(0)
    
    return jsonify({
        'observation_data': observation_data,
        'first_question': {
            'id': first_question['id'],
            'text': str(first_question['text']),
            'category': str(first_question['category']),
            'subcategory': str(first_question['subcategory']) if first_question['subcategory'] else None
        },
        'total_questions': get_total_question_count()
    }), 200


@observations_api_bp.route('/save', methods=['POST'])
@jwt_required_api
def save_observation():
    """
    POST /api/v1/observations/save
    
    Save a completed observation.
    
    Request: {
        "observation_data": {
            "session_id": 1,
            "participant_id": 1,
            "answers": {...},
            "previous_version": 0
        },
        "freeform_notes": "Optional notes"
    }
    Response: {
        "id": 1,
        "message": "Observation saved successfully",
        "version": 1
    }
    """
    user_id = int(get_jwt_identity())
    data = request.get_json()
    
    if not data:
        return jsonify({'error': 'Invalid request format'}), 400
    
    observation_data = data.get('observation_data')
    freeform_notes = data.get('freeform_notes', '')
    
    if not observation_data:
        return jsonify({'error': 'Observation data is required'}), 400
    
    record, error = ObservationService.save_observation(
        observation_data,
        freeform_notes,
        user_id
    )
    
    if error:
        return jsonify({'error': error}), 400
    
    return jsonify({
        'id': record.id,
        'message': 'Observation saved successfully',
        'version': record.version
    }), 201


@observations_api_bp.route('/<int:observation_id>', methods=['DELETE'])
@jwt_required_api
def delete_observation(observation_id):
    """
    DELETE /api/v1/observations/{observation_id}
    
    Delete an observation.
    
    Response: {"message": "Observation deleted successfully"}
    """
    user_id = int(get_jwt_identity())
    
    success, error = ObservationService.delete_observation(observation_id, user_id)
    
    if not success:
        return jsonify({'error': error or 'Failed to delete observation'}), 404
    
    return jsonify({'message': 'Observation deleted successfully'}), 200


@observations_api_bp.route('/questions', methods=['GET'])
@jwt_required_api
def get_questions():
    """
    GET /api/v1/observations/questions
    
    Get all observation questions.
    
    Response: {
        "questions": [...],
        "categories": [...],
        "answer_options": [...]
    }
    """
    questions = get_all_questions()
    
    # Convert lazy strings to regular strings for JSON serialization
    serializable_questions = []
    for q in questions:
        serializable_questions.append({
            'id': q['id'],
            'text': str(q['text']),
            'category': str(q['category']),
            'subcategory': str(q['subcategory']) if q['subcategory'] else None
        })
    
    return jsonify({
        'questions': serializable_questions,
        'categories': [str(cat) for cat in OBSERVATION_CATEGORIES],
        'answer_options': ANSWER_OPTIONS,
        'total_count': get_total_question_count()
    }), 200


@observations_api_bp.route('/questions/<int:index>', methods=['GET'])
@jwt_required_api
def get_question_by_idx(index):
    """
    GET /api/v1/observations/questions/{index}
    
    Get a specific question by index.
    
    Response: {
        "id": "question_id",
        "text": "Question text",
        "category": "Category",
        "subcategory": "Subcategory or null"
    }
    """
    if index < 0 or index >= get_total_question_count():
        return jsonify({'error': 'Invalid question index'}), 400
    
    question = get_question_by_index(index)
    
    return jsonify({
        'id': question['id'],
        'text': str(question['text']),
        'category': str(question['category']),
        'subcategory': str(question['subcategory']) if question['subcategory'] else None,
        'index': index,
        'total': get_total_question_count()
    }), 200
